[FEAT] Edit 페이지 구현

Da-Yeon-Kim opened the pull request:
## 주요 변경사항 ✔️
Edit 페이지를 구현하였습니다.
![image](https://github.com/user-attachments/assets/97d22ad1-86b3-4613-8eae-1786d4cc0f01)
![image](https://github.com/user-attachments/assets/2459942f-cd62-4a6f-82c2-acae2ccda769)

### 동작
- Edit 페이지에 접속하면 백으로부터 데이터를 받아와 기초값을 띄워줍니다. (위 사진처럼)
- 사용자는 이 기초값을 다시 선택 해제할 수 있습니다.
- 그 외 동작은 기존 JoinPage와 동일합니다.
- EditBtn을 누르면 백에게 PUT 요청을 해 수정된 데이터를 보냅니다.

### 수정된 사항
- `JoinCalendar` : 초기값을 selectedEvents에 추가하도록 수정
- `JoinFormContext` : 초기값 관련된 로직 생성 및 `meetingId`별로 Context 받던 것을 다시 수정 (필요 없더라고요...)
- `JonCalendar`에서 시간 관련된 로직을 모두 UTC 시간대가 아니라 로컬 시간대를 사용해 계산하도록 수정

### 추가한 사항
- `useGetPersonalEvents` 훅 : JoinPage에서 입력했던 데이터 백으로부터 다시 가져오는 GET
- `useUpdatePersonal` 훅 : [수정하기] 버튼 클릭 시, 데이터 제출하는 PUT
- `EditBtn` : 버튼 글자도 다르고, 사용하는 훅도 달라서 아예 새로 만들었습니다!
- `EditPage` : 훅을 이용해 기초값을 받아오고 이를 `FormProvider`를 이용해 Context에 할당합니다.
- 관련된 목 데이터

## 리뷰어에게... 🙏
하다 보니 생각보다 커밋 단위가 커졌네요.... 리뷰할 때 조금 불편해도 양해 부탁드립니다!

1. `JoinTitle`, `JoinCalendar`나 추후 생길 `JoinFood`는 EditPage에서도 공통으로 사용하려고 합니다. 다만, `EditBtn`은 상기한 내용처럼 새로 만들어 사용했습니다!

2. 시간 관련해서 로직이 좀... 복잡한데요. 이게 UTC 시간대로 자동 변환되는 걸 막고 로컬 시간대로 계산해야 정확한 시간대가 나와서 그렇습니다.... (UTC 변환되면 로컬로는 9시였던 시간이 0시로 바뀝니다)
거기다 사용자 선택 시 선택된 걸 취소하게 하려면 받은 시간을 30분 단위로 쪼개서 비교하고, 그걸 또 백엔드에 보낼 때는 다시 통합해서 보내줘야 해서.... 많은 일이 있었습니다😥

3. 지금 왜인지는 모르겠는데, 캘린더에 표시되는 게 primary-color가 아니라 green??으로 나와서 좀 살펴봐야 할 것 같습니다!

## 관련 이슈

closes #100

geongyu09 committed:
[FEAT] 그룹 캘린더 추가 (#83)

geongyu09 committed:
feat: 음식 컴포넌트 좌우 스크롤 되도록 수정 (#92)

음식 컴포넌트 좌우 스크롤 되도록 수정하였습니다

geongyu09 committed:
[FEAT] 그룹 캘린더 추가 (#94)

geongyu09 committed:
[FEAT] 주최자 페이지 일부 (라우팅, 캘린더) (#90)

* feat(useGetPermittion): 주최자 여부 확인

#89

* feat: 사용자가 해당 미팅의 호스트인 경우 leader 페이지로 리다이렉트

#89

* feat(GroupHostCalendar): 호스트 페이지의 그룹 켈린더 추가

* fix: 올바르게 이벤트가 백그라운드로 표시되지 않던 버그 수정

#89

* feat: 그룹 호스트 캘린더에 form 추가

#89

* feat(GroupSelectedTime): 선택된 시간 보여주는 컴포넌트 추가

#89"

geongyu09 committed:
[FEAT] 그룹 관리 페이지 제목 추가 (#98)

* feat: 복사 링크 아이콘 추가

* feat: 그룹 관리 페이지 타이틀 추가

Da-Yeon-Kim committed:
[FEAT] Join 버튼 로직 (#99)

* feat(JoinCalendar, JoinBtn): 버튼 로직

- useJoinMeeting.ts 훅 이용해 JoinBtn 클릭 시 폼 제출
- useJoinFormContext 이용해 폼 관리
- 이벤트는 displayedEvent와 selectedEvents로 분리해 관리
- selectedEvents에 관한 로직은 toggleSelectedEvent로 분리

 #81

* feat(WeeklyCalendar): all-Day 제거 및 스타일링 변경

- all-Day 슬롯 없게 변경
- highlight 색상 변경

 #81

* feat(JoinCalendar): 선택 불가능 판단 로직 수정

- 겹치는지 확인하는 로직을 더 세부적이게 추가
- 잘못 작성된 mock data 수정
(이것 때문에 정상 작동이 안 된 거였다....)

 #81

* refactor(JoinCalendar): isOverlapping 로직 분리

+) WeeklyCalendar에서 onSelectTime 선택적 제공하도록 수정

 #81

* feat(Join): times를 props로 받도록 변경

- Join 페이지에서 상태 관리를 하고 하위 컴포넌트에게 props로 넘겨주는
  방식으로 변경

  #81

* feat(Join): props 대신 Context API 사용

- formProvider랑 병용하려고 했는데 실패해서 일단 Context API만 단독 사용
  중... 유효성 검사는 그냥 수제로....
- Context API 이용해 times, preferences 등 전역 상태로 관리

 #81

* chore(JoinBtn): 콘솔문 삭제

 #81

* chore(Join): PR 리뷰 반영

- 무의미한 React import 삭제
- undefined 대신 null 사용

 #81

* feat(Join): Context를 meetingId에 따라 다르게

 #81

geongyu09 committed:
[FEAT] 모임 확정 날짜 추가 (#106)

HyeonChae1104 committed:
[FEAT] MyPage 프로필 조회 구현 (#96)

* feat(Header): 헤더에 프로필 아이콘 클릭 시 마이페이지로 이동 기능 추가

- handleUserIconClick 함수 추가하여 HeaderUserIcon에 onClick으로 전달
로고 클릭 시 메인 페이지로 이동하는 기능
HeaderUserIcon에 onClick 전달

#64

* feat(MyPage):  마이페이지 UI 및 프로필 조회 기능 추가

- - 백엔드에서 사용자 프로필 정보를 받아와 화면에 표시
- 프로필 이미지와 닉네임을 중앙 상단에 배치
- mypage에서 사용자 프로필 이미지를 큰 사이즈로 표시하기 위해 UserProfileImage 컴포넌트에  "lg" 사이즈 옵션을 추가

#64

* refactor(HeaderUserIcon): UI와 클릭로직의 책임 분리

- HeaderUserIcon은 UI를 담당,
- Header에서 클릭 시 마이페이지로 이동하는 것을 수행하도록 분리하였습니다.

#64

* refactor(MyPage): presentation-container 패턴 적용

- useGetUserProfile 훅을 사용해 데이터를 가져오고, MyPage는 props로 전달받아 렌더링만 담당
- MyPage 컴포넌트를 상위 MyPageContainer에서 데이터 로직을 처리하도록 리팩토링하였습니다.

#64

* refactor(HeadUserIcon): HeaderUserIcon 컴포넌트 재사용성 개선

- HeaderUserIcon에서 navigate 기능을 제거하고, 프로필 이미지를 렌더링하는 역할에 집중,
- Header 컴포넌트에서 RouterPath 객체 사용

#64

* refactor(HeaderUserIcon): onClick 제거

- onClick 제거
- Header에서 Link태그를 사용하도록 변경하였습니다.

#64

* refactor(MyPage): page 컴포넌트를 부르도록 routes 수정

- Container에서 Section으로 컴포넌트 이름 변경
- page 컴포넌트가 최상위가 되도록 수정하였습니다.

#64

* refactor(MyPage): presentation-container 패턴 적용 수정

- useGetUserProfile 사용하여 사용자 데이터 가져옵니다.
- UserProfileImage의 로직 사용
- ProfileSection 컴포넌트에서 데이터 관련 부분을 관리하고 `MyPage/index.tsx`에서는 해당 섹션을 불러와 렌더링합니다.

#64

* refactor(MyPage): 절대 경로로 import 경로 수정

#64

Da-Yeon-Kim committed:
feat(Edit): Edit 페이지 구현 및 Context 수정

- 기존 useJoinFormContext에서 기초값을 설정하도록 수정
- EditBtn, EditPage 추가
- useGetPersonalEvents: Join에서 추가한 일정 조회하기
- useUpdatePersonal: EditPage의 제출
- 그 외 라우팅, 목 데이터 추가

 #100

Da-Yeon-Kim committed:
feat(Edit): 기초값 설정하도록 변경

- JoinFormContext에 기초값 저장하도록 변경
- JoinCalendar로는 props로 전달

 #100

Da-Yeon-Kim committed:
feat(Edit): props 말고 formprovider로 times 업뎃

 #100

Da-Yeon-Kim committed:
feat(EditBtn): 훅 수정 및 목 데이터 추가

- 수정하기 제출 동작 관련 훅

 #100

Da-Yeon-Kim committed:
feat(Edit): 로컬-UTC 시간대 혼용 문제 해결

+) 불필요한 console.log문 삭제

 #100

HyeonChae1104 reviewed (approved):
pr 코멘트를 꼼꼼하게 남겨주셔서 읽어보면서 코드도 이해해보느라 조금 늦었네요.!
로컬 시간대로 계산하는 부분은 엄청 복잡하네요..
어려우셨을 부분을 엄청 잘하셔서,, 신기했습니다 👍👍 
수고많으셨습니다!!

캘린더 표시 색상은 저도 파일 찾아봤는데 이유는 못찾았습니당,, 하하ㅠ

geongyu09 commented on src/api/hooks/useGetPersonalEvents.ts:
```diff
@@ -0,0 +1,21 @@
+
+export const getPersonalEvents = async (meetingId: string): Promise<PersonalEvent[]> => {
+  const response = await fetchWithToken.get(getPersonalPath({ meetingId }));
+  return response.data.meeting_personal_times;
```
데이터에 대한 가공은 api 로직 레벨 보다는 훅 레벨에서 하는 것이 조금 더 유지보수적으로 편할 것 같습니다!
같은 주소, 같은 method로 가져온 데이터를 다른 곳에서 사용하게 된다면, api 로직도 다시 짜야하는 경우가 생길 수 있습니다!

Da-Yeon-Kim replied:
EditPage에서 `initialSelectedTimes`를 만드는 로직 말씀하시는 거죠?

고민을 조금 해봤는데,,, 훅 레벨에서 하게 되면, 데이터를 가공하지 않고 해당 훅을 사용하고 싶을 때 문제가 있을 것 같습니다. (30분 단위로 쪼개지지 않은 데이터를 원할 수 있으니까요) 훅 레벨에서는 데이터를 가져오는 역할만 하는 게 더 예쁘기도 하고요!

그래서 위치는 그대로 유지하되, 따로 `@/utils/calendar`로 로직을 분리해서 사용하는 식은 어떤지 물어보고 싶어요!

geongyu09 commented on src/hooks/useJoinFormContext.tsx:
```diff
@@ -1,4 +1,4 @@
-import { createContext, useContext, useState } from 'react';
+import React, { createContext, useContext, useEffect, useState } from 'react';
```
React 는 제거해도 괜찮아보입니다!

geongyu09 commented on src/pages/Edit/index.tsx:
```diff
@@ -0,0 +1,72 @@
+  if (
+    isMeetingInfoLoading ||
+    isPersonalEventsLoading ||
+    !meetingInfo ||
```
loading==true 상태와 !data 상태를 동시에 포함하는 것이 state == pending입니다!
이를 사용하면 조건문을 절반으로 줄일 수 있을 것 같아요!

geongyu09 reviewed (approved):
수고많으셨습니다..!
프론트에서 가장 어렵다는 Form 부분을 잘 개발하셨네요..!

Da-Yeon-Kim committed:
refactor(Edit): PR 리뷰 반영해 수정

- response.data로 받아오도록 수정
- 받은 데이터 30분 단위로 쪼개는 로직 유틸로 분리
- 불필요한 React import 제거
- isLoading 대신에 pending과 error 처리로 변경

 #100

Da-Yeon-Kim commented:
PR 리뷰 해주신 부분 수정했습니다~~😀
그 외에 30분으로 쪼개는 함수를 따로 유틸로 분리해서 사용하도록 변경했습니다!
혹시 다른 데에서도 쓰일 수 있을까 봐요~

(혹시 커멘트 알림이 다른 게 갔다면,, 모른 척해주세요! 잘못 이해해서 단 거라 다시 삭제했습니다... 하핫)

Da-Yeon-Kim reviewed (commented):
크게 변경된 부분은 없는 것 같아 머지 하려고 하는데, 저희 계속 Weekly10에 머지 하나요??

geongyu09 commented:
아 그렇네요.. 
일단 Weekly10 을 develop에 머지하고, 이후에 Weekly11을 파서 머지하면 될 것 같습니다!
pr 제가 날리도록 하겠습니다

geongyu09 commented:
Weekly11로 target branch 변경해주시면 됩니다!

Da-Yeon-Kim commented:
타겟 브랜치를 바꾸니까 이전 커밋이 같이 와버리네요! 아예 이 PR은 닫고, rebase 해서 다시 PR 날리겠습니다!
